openfl/experimental/workflow/transport/grpc/director_client.py

```diff
@@ -0,0 +1,269 @@
+# Copyright 2020-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+"""DirectorClient module."""
+
+import logging
+from pathlib import Path
+from typing import Any, Dict, Iterator, Optional, Tuple, Union  # type: ignore
+
+import grpc
+from grpc._channel import _MultiThreadedRendezvous as DataStream
+
+from openfl.experimental.workflow.protocols import director_pb2, director_pb2_grpc
+from openfl.experimental.workflow.transport.grpc.exceptions import EnvoyNotFoundError
+
+from .grpc_channel_options import channel_options
+
+logger = logging.getLogger(__name__)
+
+
+class DirectorClient:
+    """Director client class for experiment managers/envoys.
+
+    This class communicates with the director to manage the envoys
+    participation in the federation.
+
+    Attributes:
+        director_addr (host:port): Director Address
+        envoy_name (Optional[str]): The name of the envoy.
+        stub (director_pb2_grpc.DirectorStub): The gRPC stub for communication
+            with the director.
+    """
+
+    def __init__(
+        self,
+        *,
+        director_host: str,
+        director_port: int,
+        envoy_name: Optional[str] = None,
```

Comment:
Why does the `DirectorClient` need to know anything about envoys?

Reply:
Currently, a single client handles both FederatedRuntime and Envoys. 
So this will be refactored into two distinct clients: EnvoyClient and FederatedRuntimeClient.
The EnvoyClient will need information about envoys. This refactoring is a work in 
progress and will be included in a follow-up PR.